selfdrive/ui/sound.cc

```diff
@@ -1,91 +1,119 @@
-  { cereal::CarControl::HUDControl::AudibleAlert::NONE, NULL, false },
-};
-
-sound_file* get_sound_file(AudibleAlert alert) {
-  for (sound_file *s = sound_table; s->alert != cereal::CarControl::HUDControl::AudibleAlert::NONE; s++) {
-    if (s->alert == alert) {
-      return s;
-    }
+void SLAPIENTRY slplay_callback(SLPlayItf playItf, void* context, SLuint32 event) {
+  int *repeat  = (int*)context;
+  if (event == SL_PLAYEVENT_HEADATEND && --(*repeat) > 0) {
+    (*playItf)->SetPlayState(playItf, SL_PLAYSTATE_STOPPED);
+    (*playItf)->SetMarkerPosition(playItf, 0);
+    (*playItf)->SetPlayState(playItf, SL_PLAYSTATE_PLAYING);
   }
-
-  return NULL;
 }
 
-void play_alert_sound(AudibleAlert alert) {
-  sound_file* sound = get_sound_file(alert);
-  char* error = NULL;
+bool Sound::play(AudibleAlert alert, int repeat) {
+  stop();
+  currentSound_ = alert;
+  auto playerItf = player_.at(alert)->playInterface;
+  if (repeat > 0) {
+    repeat_ = repeat;
+    ReturnOnError((*playerItf)->RegisterCallback(playerItf, slplay_callback, &repeat_), "Failed to register callback");
+    ReturnOnError((*playerItf)->SetCallbackEventsMask(playerItf, SL_PLAYEVENT_HEADATEND), "Failed to set callback event mask");
+  }
 
-  slplay_play(sound->uri, sound->loop, &error);
-  if(error) {
-    LOGW("error playing sound: %s", error);
+  // Reset the audio player
+  ReturnOnError((*playerItf)->ClearMarkerPosition(playerItf), "Failed to clear marker position");
+  uint32_t states[] = {SL_PLAYSTATE_PAUSED, SL_PLAYSTATE_STOPPED, SL_PLAYSTATE_PLAYING};
+  for (auto state : states) {
+    ReturnOnError((*playerItf)->SetPlayState(playerItf, state), "Failed to set SL_PLAYSTATE_PLAYING");
   }
+  return true;
 }
 
-void stop_alert_sound(AudibleAlert alert) {
-  sound_file* sound = get_sound_file(alert);
-  char* error = NULL;
-
-  slplay_stop_uri(sound->uri, &error);
-  if(error) {
-    LOGW("error stopping sound: %s", error);
+bool Sound::stop() {
+  repeat_ = 0;
+  if (currentSound_ != AudibleAlert::NONE) {
+    auto playerItf = player_.at(currentSound_)->playInterface;
+    currentSound_ = AudibleAlert::NONE;
+    ReturnOnError((*playerItf)->SetPlayState(playerItf, SL_PLAYSTATE_PAUSED), "Failed to set SL_PLAYSTATE_STOPPED");
   }
+  return true;
 }
 
-void ui_sound_init() {
-  char *error = NULL;
-  slplay_setup(&error);
-  if (error) goto fail;
-
-  for (sound_file *s = sound_table; s->alert != cereal::CarControl::HUDControl::AudibleAlert::NONE; s++) {
-    slplay_create_player_for_uri(s->uri, &error);
-    if (error) goto fail;
+void Sound::setVolume(int volume, int timeout_seconds) {
+  if (last_volume_ == volume) {
+    return;
+  }
+  double current_time = nanos_since_boot();
+  if ((current_time - last_set_volume_time_) > (timeout_seconds * (1e+9))) {
+    char volume_change_cmd[64];
+    snprintf(volume_change_cmd, sizeof(volume_change_cmd), "service call audio 3 i32 3 i32 %d i32 1 &", volume);
+    system(volume_change_cmd);
```

Comment:
this shouldn't run on PC

Reply:
sound.cc will not be compiled on PC. there is a dummy function setVolume in linux.cc for PC